### PR TITLE
WIP: Tree folders view

### PR DESCRIPTION
**Pull request by edoardocavazza**
This is a tracking PR that introduces a new tree folders navigation as described in #397.

- [x] Dynamically load children
- [x] Styling
- [x] Improve routing
- [x] Handle children pagination
- [x] Restore parents in tree
- [x] Save object position
- [x] Save folder position
- [x] Add tests

### CODE REVIEW

File: src/Controller/ModulesController.php [hunk 1]
```diff
@@ -458,6 +462,44 @@ public function relatedJson($id, string $relation): void
...
+                $query['filter']['roots'] = '';
+                $response = $this->apiClient->getObjects('folders', $query);
+            } else {
+                if (empty($query['filter'])) {
+                    $query['filter'] = [];
+                }
+                $query['filter']['parent'] = $root;
+                $response = $this->apiClient->getObjects('folders', $query);
+            }
+        } catch (BEditaClientException $error) {
```
Comment (didoda) [inline src/Controller/ModulesController.php:489]: you could simplify code [lines 477-489]; i.e.:

```
$key = (empty($root)) ? 'roots' : 'parent';
$val = (empty($root)) ? '' : $root;
if (empty($query['filter'])) {
    $query['filter'] = [];
}
$query['filter'][$key] = $val;
$response = $this->apiClient->getObjects('folders', $query);
```

File: src/Controller/ModulesController.php [hunk 2]
```diff
@@ -458,6 +462,38 @@ public function relatedJson($id, string $relation): void
...
+     * @param string|int $root the tree root identifier.
+     * @return void
+     */
+    public function treeJson($id, string $root = null): void
+    {
+        $this->request->allowMethod(['get']);
+        $query = $this->Modules->prepareQuery($this->request->getQueryParams());
+        try {
+            $key = (empty($root)) ? 'roots' : 'parent';
+            $val = (empty($root)) ? 'roots' : $root;
```
Comment (didoda) [inline src/Controller/ModulesController.php]: `$val = (empty($root)) ? '' : $root;`

File: config/routes.php [hunk 3]
```diff
@@ -200,9 +200,14 @@
         ['pass' => ['id', 'relation'], '_name' => 'modules:resourcesJson']
     );
     $routes->connect(
-        '/:object_type/view/:id/treeJson/:relation',
+        '/:object_type/view/:id/treeJson',
```
Comment (stefanorosanelli) [inline config/routes.php:203]: you could keep only the first rule and check `root` as optional query string in `treeJson` method

[depending on how you generate routes...]

File: src/Controller/AppController.php [hunk 4]
```diff
@@ -182,7 +182,11 @@ protected function prepareRequest($type): array
                 $id = $data['id'];
 
                 foreach ($relationData as $method => $ids) {
-                    $relatedIds = json_decode($ids, true);
+                    $relatedIds = is_string($ids) ? json_decode($ids, true) : array_map(
```
Comment (stefanorosanelli) [inline src/Controller/AppController.php]: In order to ease debugging we avoid ternary `?` operator usually... 

```php
     if (is_string($ds)) {
           $relatedIds = is_string($ids);
     } else { 
           $relatedIds = array_map(...);
     }
```

but you can keep it like this

**Review by stefanorosanelli — APPROVED** [review]

File: config/routes.php [hunk 5]
```diff
@@ -200,9 +200,9 @@
...
         ['controller' => 'Modules', 'action' => 'treeJson'],
-        ['pass' => ['id', 'relation'], '_name' => 'modules:treeJson']
+        ['pass' => ['id', 'root'], '_name' => 'modules:treeJson']
```
Comment (stefanorosanelli) [inline config/routes.php]: we can remove `root' from `pass` array now since is not passed as argument to `treeJson` method anymore

**Pull request merged**